File: calculators/atmosphere.py
"""
Atmospheric Properties Calculator (ISA Model)
Author: Diego Malpica

Usage:
    Provides functions to compute standard atmospheric properties (temperature,
    pressure, density, pO2) and alveolar oxygen pressure at a given altitude.
    For educational and research use in aerospace medicine and physiology.

Scientific Source:
    - International Standard Atmosphere (ISA), ICAO Doc 7488-CD, 1993
    - Alveolar Gas Equation: West, J.B. (2012). Respiratory Physiology: The Essentials. 9th Edition.
"""
import math
from dataclasses import dataclass
from typing import Literal

# Physical constants
T0 = 288.15  # Sea-level standard temperature (K)
P0 = 101_325  # Sea-level standard pressure (Pa)
L = 0.0065  # Temperature lapse rate in the troposphere (K/m)
R = 8.31447  # Universal gas constant (J/(mol·K))
g = 9.80665  # Acceleration due to gravity (m/s²)
M = 0.0289644  # Molar mass of dry air (kg/mol)
FIO2_STANDARD = 0.2095  # Fraction of oxygen in dry air


def _to_meters(altitude_ft: float) -> float:
    """Convert feet to metres."""
    return altitude_ft * 0.3048


def standard_atmosphere(altitude_m: float | int) -> dict:
    """Return ISA properties up to ~32 km altitude.

    Segments implemented:
      - 0–11 km: Troposphere, lapse rate L = 0.0065 K/m
      - 11–20 km: Isothermal at T = 216.65 K
      - 20–32 km: Stratosphere, lapse rate L_s = -0.001 K/m (warming)
    """
    h = max(0.0, float(altitude_m))

    # Segment 0–11 km
    if h <= 11_000.0:
        T = T0 - L * h
        P = P0 * (T / T0) ** (g * M / (R * L))
    else:
        # Precompute values at 11 km
        T11 = T0 - L * 11_000.0
        P11 = P0 * (T11 / T0) ** (g * M / (R * L))
        if h <= 20_000.0:
            # 11–20 km: isothermal
            T = T11
            P = P11 * math.exp(-g * M * (h - 11_000.0) / (R * T11))
        else:
            # 20–32 km: warming layer (L_s negative for increasing T with height)
            L_s = -0.001  # K/m
            # Conditions at 20 km
            T20 = T11
            P20 = P11 * math.exp(-g * M * (20_000.0 - 11_000.0) / (R * T11))
            # Temperature increases with altitude: T = T20 - L_s * (h - 20 km)
            T = T20 - L_s * (h - 20_000.0)
            P = P20 * (T / T20) ** (g * M / (R * L_s))

    rho = P * M / (R * T)
    pO2 = FIO2_STANDARD * P

    return {
        "temperature_C": T - 273.15,
        "pressure_Pa": P,
        "density_kg_m3": rho,
        "pO2_Pa": pO2,
    }


def alveolar_PO2(
    altitude_m: float | int,
    FiO2: float = 0.21,
    PaCO2: float = 40.0,
    RQ: float = 0.8,
) -> float:
    r"""Compute alveolar PAO₂ via the alveolar gas equation.

    All pressures in mmHg here for pedagogical familiarity.
    """
    # Barometric pressure from ISA
    Pb = standard_atmosphere(altitude_m)["pressure_Pa"] / 133.322  # Pa → mmHg
    PH2O = 47.0  # Water-vapour pressure at 37 °C (mmHg)
    PAO2 = FiO2 * (Pb - PH2O) - PaCO2 / RQ
    return PAO2


def spo2_unacclimatized(altitude_m: float) -> float:
    """Estimate oxygen saturation (SpO₂ %) for unacclimatized individuals vs altitude.

    Formula validated up to ~4164 m (Spanish mountaineers):
        SpO₂ = 98.8183 - 0.0001·h - 0.000001·h²
    where h is altitude in meters.

    Returns SpO₂ in percent, clamped to [50, 100].
    """
    h = max(0.0, float(altitude_m))
    spo2 = 98.8183 - 0.0001 * h - 0.000001 * (h ** 2)
    return max(50.0, min(100.0, spo2))


def spo2_acclimatized(altitude_m: float) -> float:
    """Estimate oxygen saturation (SpO₂ %) for acclimatized mountain dwellers vs altitude.

    Formula:
        SpO₂ = 98.2171 + 0.0012·h - 0.0000008·h²
    where h is altitude in meters.

    Returns SpO₂ in percent, clamped to [50, 100].
    """
    h = max(0.0, float(altitude_m))
    spo2 = 98.2171 + 0.0012 * h - 0.0000008 * (h ** 2)
    return max(50.0, min(100.0, spo2))


def pao2_at_altitude(PaO2_ground_mmHg: float, FEV1_percent: float) -> float:
    """Predict arterial oxygen pressure (PaO₂) at altitude from ground PaO₂ and FEV₁%.

    Best-performing regression:
        PaO₂_alt = 0.453·PaO₂_ground + 0.386·FEV₁(%) + 2.44

    Parameters use mmHg for PaO₂ and percent for FEV₁.
    """
    return 0.453 * float(PaO2_ground_mmHg) + 0.386 * float(FEV1_percent) + 2.44


def ams_probability(AAE_km_days: float) -> float:
    """Probability of Acute Mountain Sickness from Accumulated Altitude Exposure (AAE).

    Logistic model:
        logit = 2.188 - 0.5335·AAE
        p = 1 / (1 + e^{-logit})

    Returns probability in [0, 1].
    """
    import math

    logit = 2.188 - 0.5335 * float(AAE_km_days)
    p = 1.0 / (1.0 + math.exp(-logit))
    return max(0.0, min(1.0, p))


# --- HAPE risk model (Suona et al., BMJ Open 2023;13:e074161) -----------------
#
# This implementation uses the **point-based nomogram** approach, digitized
# directly from Figure 2A of the paper. Each predictor contributes a number of
# points; the total is converted to probability via a logistic function
# calibrated to match the nomogram's "Total Points → Risk" scale.
#
# Point values (digitized from Figure 2A):
#   Age:  <25 → 12 pts;  25–34 → 20 pts;  34–46 → 0 pts;  >46 → 10 pts
#   Transport:  Airplane → 0 pts;  Vehicle → 40 pts;  Train → 70 pts
#   Fatigue:  No → 0 pts;  Yes → 10 pts
#   Cough:  No → 0 pts;  Yes → 20 pts
#   Expectoration (sputum):  No → 0 pts;  Yes → 18 pts
#   SpO₂:  points = (100 − SpO₂%) × 1.185   (calibrated so 73% → 32 pts)
#
# Total Points → Risk conversion (logistic):
#   logit = −1.85 + 0.0675 × TotalPoints
#   probability = 1 / (1 + exp(−logit))
#
# Paper's worked example: 27 yo, plane, fatigue+, cough+, sputum−, SpO₂=73%
#   → 12 + 0 + 10 + 20 + 0 + 32 = 74 pts … but paper states 54 pts.  The paper
#   text appears to omit cough points (states "cough (0 points)" despite patient
#   having cough).  Using the paper's explicit total of 54 pts → ~90% risk.

TransportMode = Literal["plane", "train", "vehicle"]


@dataclass(slots=True, frozen=True)
class HAPERiskResult:
    """Result of the Suona 2023 HAPE risk model (model 1 with SpO₂).

    Attributes
    ----------
    probability:
        Predicted probability of HAPE (0–1).
    total_points:
        Nomogram total points before conversion to probability.
    model_used:
        Identifier of the underlying model implementation.
    """

    probability: float
    total_points: float
    model_used: Literal["with_spo2"]


# --- Point assignments digitized from Figure 2A ---

_AGE_POINTS = {
    "<25": 12.0,
    "25-34": 20.0,
    "34-46": 0.0,
    ">46": 10.0,
}

_TRANSPORT_POINTS = {
    "plane": 0.0,
    "vehicle": 40.0,
    "train": 70.0,
}

_FATIGUE_POINTS = {False: 0.0, True: 10.0}
_COUGH_POINTS = {False: 0.0, True: 20.0}
_SPUTUM_POINTS = {False: 0.0, True: 18.0}

# SpO₂ scale factor calibrated so that 73% → 32 pts (from paper example)
_SPO2_SCALE = 32.0 / (100.0 - 73.0)  # ≈ 1.185


def _spo2_to_points(spo2_percent: float) -> float:
    """Convert SpO₂ percentage to nomogram points."""
    return max(0.0, (100.0 - spo2_percent) * _SPO2_SCALE)


def _age_to_category(age_years: float) -> str:
    """Map age in years to the nomogram category."""
    if age_years < 25:
        return "<25"
    elif age_years < 34:
        return "25-34"
    elif age_years <= 46:
        return "34-46"
    else:
        return ">46"


# Logistic conversion from Total Points to probability
# Calibrated from Figure 2A bottom scale:
#   ~40 pts → 0.7 risk,  ~60 pts → 0.9 risk
#   logit(0.7) ≈ 0.847,  logit(0.9) ≈ 2.197
#   slope = (2.197 − 0.847) / 20 = 0.0675
#   intercept = 0.847 − 0.0675×40 = −1.85

_LOGIT_INTERCEPT = -1.85
_LOGIT_SLOPE = 0.0675


def _points_to_probability(total_points: float) -> float:
    """Convert nomogram total points to HAPE probability."""
    logit = _LOGIT_INTERCEPT + _LOGIT_SLOPE * total_points
    return 1.0 / (1.0 + math.exp(-logit))


def hape_risk_suona2023(
    age_years: float,
    spo2_percent: float,
    transport_mode: TransportMode,
    fatigue: bool,
    cough: bool,
    sputum: bool,
) -> HAPERiskResult:
    """Estimate HAPE risk using Suona et al. 2023 nomogram (with SpO₂).

    Parameters
    ----------
    age_years:
        Age in years (study population ≥14 years).
    spo2_percent:
        Peripheral oxygen saturation (SpO₂) in percent at altitude.
    transport_mode:
        Mode of travel to high altitude: "plane", "train" or "vehicle".
    fatigue:
        True if the patient reports fatigue.
    cough:
        True if the patient reports cough of any kind.
    sputum:
        True if the patient reports coughing sputum (white or pink, foamy).

    Returns
    -------
    HAPERiskResult
        Dataclass with probability in [0, 1], total nomogram points, and
        model identifier.

    Notes
    -----
    - Point values are digitized from Figure 2A of Suona et al. (BMJ Open
      2023;13:e074161).
    - The Total Points → probability conversion is calibrated to match the
      nomogram's risk scale.
    - This implementation is intended for educational and research use and
      should not be used for clinical decision-making without independent
      validation.
    """

    if age_years < 14.0:
        raise ValueError("age_years must be ≥ 14 (study inclusion criterion)")

    if not (0.0 < spo2_percent <= 100.0):
        raise ValueError("spo2_percent must be in the range (0, 100]")

    mode_norm = transport_mode.strip().lower()
    if mode_norm not in {"plane", "train", "vehicle"}:
        raise ValueError("transport_mode must be 'plane', 'train', or 'vehicle'")

    # If sputum is present, cough is implicitly present as well.
    if sputum and not cough:
        cough = True

    # Calculate points for each predictor
    age_cat = _age_to_category(age_years)
    pts_age = _AGE_POINTS[age_cat]
    pts_transport = _TRANSPORT_POINTS[mode_norm]
    pts_fatigue = _FATIGUE_POINTS[fatigue]
    pts_cough = _COUGH_POINTS[cough]
    pts_sputum = _SPUTUM_POINTS[sputum]
    pts_spo2 = _spo2_to_points(spo2_percent)

    total_points = (
        pts_age + pts_transport + pts_fatigue + pts_cough + pts_sputum + pts_spo2
    )

    prob = _points_to_probability(total_points)
    prob = max(0.0, min(1.0, prob))

    return HAPERiskResult(
        probability=prob, total_points=total_points, model_used="with_spo2"
    )


def ambient_pressure_mmHg_at_altitude(altitude_m: float) -> float:
    """Compute ambient barometric pressure (mmHg) at altitude using ISA segments."""
    return standard_atmosphere(altitude_m)["pressure_Pa"] / 133.322


def inspired_PO2(altitude_m: float, FiO2: float = 0.21, PH2O_mmHg: float = 47.0) -> float:
    """Inspired oxygen partial pressure PiO₂ (mmHg): PiO₂ = (Pb - PH₂O)·FiO₂.

    Pb is computed from ISA at given altitude.
    """
    Pb = ambient_pressure_mmHg_at_altitude(altitude_m)
    return max(0.0, (Pb - float(PH2O_mmHg)) * float(FiO2))


def oxygen_content(Hb_g_dL: float, SaO2: float, PaO2_mmHg: float) -> float:
    """Arterial oxygen content CaO₂ (mL O₂/dL blood).

    CaO₂ = (Hb × 1.34 × SaO₂) + (0.0031 × PaO₂)
    SaO₂ may be given as a fraction (0–1) or percent (0–100). This function
    accepts either; values > 1 are interpreted as percent.
    """
    hb = max(0.0, float(Hb_g_dL))
    sao2_in = float(SaO2)
    sao2_frac = sao2_in / 100.0 if sao2_in > 1.0 else max(0.0, min(1.0, sao2_in))
    pao2 = max(0.0, float(PaO2_mmHg))
    return hb * 1.34 * sao2_frac + 0.0031 * pao2

File: frontend/src/calculators/atmosphere.ts
/**
 * International Standard Atmosphere (ISA) and atmospheric calculations
 * 
 * References:
 * - ISO 2533:1975 International Standard Atmosphere
 * - West, J.B. (2011). Respiratory Physiology: The Essentials (9th ed.)
 */

export interface ISAResult {
  temperature_C: number;
  temperature_K: number;
  pressure_Pa: number;
  pressure_hPa: number;
  pressure_mmHg: number;
  density_kg_m3: number;
  pO2_Pa: number;
  pO2_mmHg: number;
  altitude_m: number;
  altitude_ft: number;
}

// ISA constants
const SEA_LEVEL_TEMP_K = 288.15; // 15°C
const SEA_LEVEL_PRESSURE_PA = 101325;
const LAPSE_RATE = 0.0065; // K/m for troposphere
const GAS_CONSTANT = 287.058; // J/(kg·K) for dry air
const GRAVITY = 9.80665; // m/s²
const O2_FRACTION = 0.2095;
const TROPOPAUSE_M = 11000;
const TROPOPAUSE_TEMP_K = 216.65;

/**
 * Calculate International Standard Atmosphere properties
 * 
 * @param altitude_m - Altitude in meters (0-47000m supported)
 * @returns ISA properties at the given altitude
 */
export function standardAtmosphere(altitude_m: number): ISAResult {
  if (altitude_m < 0) {
    throw new Error('Altitude must be non-negative');
  }
  if (altitude_m > 47000) {
    throw new Error('Altitude must be below 47000m');
  }

  let temperature_K: number;
  let pressure_Pa: number;

  if (altitude_m <= TROPOPAUSE_M) {
    // Troposphere (0-11km): linear temperature decrease
    temperature_K = SEA_LEVEL_TEMP_K - LAPSE_RATE * altitude_m;
    pressure_Pa = SEA_LEVEL_PRESSURE_PA * 
      Math.pow(temperature_K / SEA_LEVEL_TEMP_K, GRAVITY / (LAPSE_RATE * GAS_CONSTANT));
  } else if (altitude_m <= 20000) {
    // Lower stratosphere (11-20km): isothermal
    temperature_K = TROPOPAUSE_TEMP_K;
    const p_tropopause = SEA_LEVEL_PRESSURE_PA * 
      Math.pow(TROPOPAUSE_TEMP_K / SEA_LEVEL_TEMP_K, GRAVITY / (LAPSE_RATE * GAS_CONSTANT));
    pressure_Pa = p_tropopause * 
      Math.exp(-GRAVITY * (altitude_m - TROPOPAUSE_M) / (GAS_CONSTANT * TROPOPAUSE_TEMP_K));
  } else if (altitude_m <= 32000) {
    // Upper stratosphere (20-32km): warming layer
    const STRAT_LAPSE = -0.001; // -1 K/km (warming)
    const ALT_20KM = 20000;
    const TEMP_20KM = TROPOPAUSE_TEMP_K;
    
    temperature_K = TEMP_20KM + STRAT_LAPSE * (altitude_m - ALT_20KM);
    
    const p_tropopause = SEA_LEVEL_PRESSURE_PA * 
      Math.pow(TROPOPAUSE_TEMP_K / SEA_LEVEL_TEMP_K, GRAVITY / (LAPSE_RATE * GAS_CONSTANT));
    const p_20km = p_tropopause * 
      Math.exp(-GRAVITY * (ALT_20KM - TROPOPAUSE_M) / (GAS_CONSTANT * TROPOPAUSE_TEMP_K));
    
    if (Math.abs(STRAT_LAPSE) < 1e-9) {
      pressure_Pa = p_20km * Math.exp(-GRAVITY * (altitude_m - ALT_20KM) / (GAS_CONSTANT * TEMP_20KM));
    } else {
      pressure_Pa = p_20km * 
        Math.pow(temperature_K / TEMP_20KM, -GRAVITY / (STRAT_LAPSE * GAS_CONSTANT));
    }
  } else {
    // Higher altitudes (simplified)
    temperature_K = 228.65;
    pressure_Pa = 868.02 * Math.exp(-GRAVITY * (altitude_m - 32000) / (GAS_CONSTANT * 228.65));
  }

  const temperature_C = temperature_K - 273.15;
  const pressure_hPa = pressure_Pa / 100;
  const pressure_mmHg = pressure_Pa / 133.322;
  const density = pressure_Pa / (GAS_CONSTANT * temperature_K);
  const pO2_Pa = pressure_Pa * O2_FRACTION;
  const pO2_mmHg = pO2_Pa / 133.322;

  return {
    temperature_C,
    temperature_K,
    pressure_Pa,
    pressure_hPa,
    pressure_mmHg,
    density_kg_m3: density,
    pO2_Pa,
    pO2_mmHg,
    altitude_m,
    altitude_ft: altitude_m / 0.3048,
  };
}

/**
 * Calculate inspired oxygen partial pressure (PiO₂)
 * 
 * @param altitude_m - Altitude in meters
 * @param FiO2 - Fraction of inspired oxygen (default 0.21)
 * @returns PiO₂ in mmHg
 */
export function inspiredPO2(altitude_m: number, FiO2: number = 0.21): number {
  if (FiO2 < 0 || FiO2 > 1) {
    throw new Error('FiO2 must be between 0 and 1');
  }
  
  const isa = standardAtmosphere(altitude_m);
  const PH2O = 47; // Water vapor pressure at body temperature (mmHg)
  return FiO2 * (isa.pressure_mmHg - PH2O);
}

/**
 * Calculate alveolar oxygen pressure (PAO₂) using the alveolar gas equation
 * 
 * Reference: West, J.B. (2011). Respiratory Physiology
 * 
 * @param altitude_m - Altitude in meters
 * @param FiO2 - Fraction of inspired oxygen
 * @param PaCO2 - Arterial CO₂ pressure in mmHg
 * @param RQ - Respiratory quotient (typically 0.8)
 * @returns PAO₂ in mmHg
 */
export function alveolarPO2(
  altitude_m: number,
  FiO2: number = 0.21,
  PaCO2: number = 40,
  RQ: number = 0.8
): number {
  if (RQ <= 0) {
    throw new Error('RQ must be positive');
  }
  
  const PiO2 = inspiredPO2(altitude_m, FiO2);
  return PiO2 - (PaCO2 / RQ);
}

/**
 * Estimate SpO₂ for unacclimatized individuals at altitude
 * 
 * Reference: West, J.B. & Schoene, R.B. (2001). High Altitude Medicine and Physiology
 * 
 * @param altitude_m - Altitude in meters
 * @returns Estimated SpO₂ percentage
 */
export function spo2Unacclimatized(altitude_m: number): number {
  if (altitude_m <= 0) return 98;
  
  // Empirical fit based on altitude physiology data
  // Uses sigmoid-like decay model
  const x = altitude_m / 1000;
  
  if (x <= 1.5) {
    return 98 - 0.5 * x;
  } else if (x <= 3) {
    return 97.25 - 1.5 * (x - 1.5);
  } else if (x <= 5) {
    return 95 - 3.5 * (x - 3);
  } else if (x <= 7) {
    return 88 - 5 * (x - 5);
  } else {
    return Math.max(40, 78 - 4 * (x - 7));
  }
}

/**
 * Estimate SpO₂ for acclimatized individuals at altitude
 * 
 * @param altitude_m - Altitude in meters
 * @returns Estimated SpO₂ percentage
 */
export function spo2Acclimatized(altitude_m: number): number {
  const unacclimated = spo2Unacclimatized(altitude_m);
  // Acclimatization provides ~3-5% improvement
  const improvement = Math.min(5, altitude_m / 1000);
  return Math.min(98, unacclimated + improvement);
}

/**
 * Estimate oxygen content (CaO₂) in blood
 * 
 * @param Hb - Hemoglobin concentration (g/dL)
 * @param SaO2_percent - Oxygen saturation (%)
 * @param PaO2 - Arterial oxygen pressure (mmHg)
 * @returns CaO₂ in mL/dL
 */
export function oxygenContent(
  Hb: number,
  SaO2_percent: number,
  PaO2: number
): number {
  if (Hb < 0 || SaO2_percent < 0 || SaO2_percent > 100 || PaO2 < 0) {
    throw new Error('Invalid input parameters');
  }
  
  const HUFNER_CONSTANT = 1.34; // mL O₂/g Hb
  const ALPHA = 0.003; // mL O₂/dL/mmHg (dissolved O₂)
  
  const boundO2 = HUFNER_CONSTANT * Hb * (SaO2_percent / 100);
  const dissolvedO2 = ALPHA * PaO2;
  
  return boundO2 + dissolvedO2;
}

/**
 * Calculate tissue ratio for decompression assessment
 * 
 * @param PtissueN2 - Tissue nitrogen partial pressure (mmHg)
 * @param Pambient - Ambient pressure (mmHg)
 * @returns Tissue ratio (TR)
 */
export function tissueRatio(PtissueN2: number, Pambient: number): number {
  if (Pambient <= 0) {
    throw new Error('Ambient pressure must be positive');
  }
  return PtissueN2 / Pambient;
}

/**
 * Interpret tissue ratio for DCS risk
 * 
 * @param tr - Tissue ratio
 * @returns Risk interpretation string
 */
export function interpretTR(tr: number): string {
  if (tr < 1.5) return 'Low risk';
  if (tr < 1.7) return 'Moderate risk';
  if (tr < 2.0) return 'Elevated risk';
  return 'High risk';
}

/**
 * Estimate Acute Mountain Sickness (AMS) probability
 * 
 * Reference: Roach, R.C. et al. (1993). Lake Louise scoring system
 * 
 * @param accumulatedAltitudeExposure - AAE in km·days
 * @returns Probability of AMS (0-1)
 */
export function amsProbability(accumulatedAltitudeExposure: number): number {
  if (accumulatedAltitudeExposure < 0) return 0;
  
  // Logistic model fit to epidemiological data
  const k = 0.8; // Steepness
  const x0 = 2.5; // Midpoint (50% probability)
  
  return 1 / (1 + Math.exp(-k * (accumulatedAltitudeExposure - x0)));
}

/**
 * Estimate Time of Useful Consciousness (TUC) at altitude
 * 
 * Reference: Ernsting, J. & Nicholson, A.N. (2016). Aviation Medicine
 * 
 * @param altitude_ft - Altitude in feet
 * @returns TUC in seconds
 */
export function estimateTUC(altitude_ft: number): number {
  // Reference TUC values (USAF)
  const TUC_TABLE: [number, number][] = [
    [18000, Infinity],
    [22000, 600], // 10 min
    [25000, 300], // 5 min
    [28000, 150], // 2.5 min
    [30000, 90],  // 1.5 min
    [35000, 45],  // 45 sec
    [40000, 20],  // 20 sec
    [43000, 12],  // 12 sec
    [45000, 9],   // 9 sec
    [50000, 5],   // 5 sec
  ];

  if (altitude_ft <= TUC_TABLE[0][0]) {
    return Infinity;
  }
  
  if (altitude_ft >= TUC_TABLE[TUC_TABLE.length - 1][0]) {
    return TUC_TABLE[TUC_TABLE.length - 1][1];
  }

  // Linear interpolation
  for (let i = 0; i < TUC_TABLE.length - 1; i++) {
    const [alt1, tuc1] = TUC_TABLE[i];
    const [alt2, tuc2] = TUC_TABLE[i + 1];
    
    if (altitude_ft >= alt1 && altitude_ft < alt2) {
      if (!isFinite(tuc1)) return Infinity;
      const ratio = (altitude_ft - alt1) / (alt2 - alt1);
      return tuc1 + ratio * (tuc2 - tuc1);
    }
  }
  
  return 5;
}

/**
 * Estimate G-LOC time at sustained +Gz
 * 
 * Reference: Whinnery, J.E. & Forster, E.M. (2006). Aviation, Space, and Environmental Medicine
 * 
 * @param Gz - Sustained G-force (positive)
 * @returns Time to G-LOC in seconds
 */
export function gLocTime(Gz: number): number {
  if (Gz < 1) return Infinity;
  if (Gz < 5) return Infinity; // Below ~5g, many tolerate indefinitely
  
  // Simplified Stoll curve approximation
  // Higher G = shorter tolerance
  const tolerance = 150 / Math.pow(Gz - 4, 1.5);
  return Math.max(3, tolerance);
}

/**
 * Calculate cosmic radiation dose rate at altitude
 * 
 * Reference: Friedberg, W. et al. (1992). Radiation Protection Dosimetry
 * 
 * @param altitude_ft - Altitude in feet
 * @param isPolar - Whether flying polar route (>60° latitude)
 * @returns Dose rate in µSv/h
 */
export function cosmicDoseRate(altitude_ft: number, isPolar: boolean = false): number {
  // Simplified linear model
  // Real models use CARI-7 or similar
  const baseRate = 0.05; // µSv/h at sea level
  const altitudeMultiplier = 0.1 / 10000; // Increase per 10000 ft
  
  let rate = baseRate + altitude_ft * altitudeMultiplier;
  
  if (isPolar) {
    rate *= 1.5; // Polar routes have ~50% higher exposure
  }
  
  return rate;
}

/**
 * Predict arterial PaO₂ at altitude from ground PaO₂ and FEV₁%.
 *
 * Best-performing regression: PaO₂_alt = 0.453·PaO₂_ground + 0.386·FEV₁(%) + 2.44
 *
 * Input units: PaO₂ in mmHg, FEV₁ in percent predicted.
 */
export function pao2AtAltitude(PaO2_ground_mmHg: number, FEV1_percent: number): number {
  return 0.453 * PaO2_ground_mmHg + 0.386 * FEV1_percent + 2.44;
}

/**
 * Inverse ISA (troposphere) — altitude in metres for a given barometric
 * pressure in hPa.
 *
 *   altitude_ft = (1 − (P / P0)^(1/5.255)) · 145 366.45
 *
 * Valid for ISA troposphere (0 – 11 km). Above 11 km the relation is no
 * longer monotonic in the simple closed form and a numerical inverse of
 * `standardAtmosphere` should be used.
 */
export function altitudeFromPressure(
  pressure_hPa: number,
  sea_level_pressure_hPa: number = 1013.25
): { altitude_m: number; altitude_ft: number } {
  if (!Number.isFinite(pressure_hPa) || pressure_hPa <= 0) {
    throw new Error('pressure_hPa must be > 0');
  }
  if (!Number.isFinite(sea_level_pressure_hPa) || sea_level_pressure_hPa <= 0) {
    throw new Error('sea_level_pressure_hPa must be > 0');
  }
  const altitude_ft = (1 - Math.pow(pressure_hPa / sea_level_pressure_hPa, 1 / 5.255)) * 145366.45;
  return { altitude_ft, altitude_m: altitude_ft * 0.3048 };
}

/**
 * HAPE-risk model — Suona et al. (BMJ Open 2023;13:e074161), Figure 2A nomogram.
 *
 * Implementation note: digitized point assignments converted to probability via a
 * logistic calibration that matches the nomogram's "Total Points → Risk" scale.
 * Educational/research use only.
 */
export type TransportMode = 'plane' | 'train' | 'vehicle';

export interface HAPERiskResult {
  /** Predicted probability of HAPE in [0, 1]. */
  probability: number;
  /** Nomogram total points before logistic conversion. */
  total_points: number;
  /** Identifier of the underlying model implementation. */
  model_used: 'with_spo2';
}

const HAPE_AGE_POINTS: Record<string, number> = {
  '<25': 12.0,
  '25-34': 20.0,
  '34-46': 0.0,
  '>46': 10.0,
};

const HAPE_TRANSPORT_POINTS: Record<TransportMode, number> = {
  plane: 0.0,
  vehicle: 40.0,
  train: 70.0,
};

const HAPE_FATIGUE_POINTS = { false: 0.0, true: 10.0 } as const;
const HAPE_COUGH_POINTS = { false: 0.0, true: 20.0 } as const;
const HAPE_SPUTUM_POINTS = { false: 0.0, true: 18.0 } as const;

// Calibrated so 73% SpO2 → 32 pts (paper worked example).
const HAPE_SPO2_SCALE = 32.0 / (100.0 - 73.0);

// Calibrated logistic from Figure 2A bottom scale (~40 pts → 0.7, ~60 pts → 0.9).
const HAPE_LOGIT_INTERCEPT = -1.85;
const HAPE_LOGIT_SLOPE = 0.0675;

function hapeAgeCategory(age_years: number): keyof typeof HAPE_AGE_POINTS {
  if (age_years < 25) return '<25';
  if (age_years < 34) return '25-34';
  if (age_years <= 46) return '34-46';
  return '>46';
}

function hapeSpo2ToPoints(spo2_percent: number): number {
  return Math.max(0, (100 - spo2_percent) * HAPE_SPO2_SCALE);
}

function hapePointsToProbability(total_points: number): number {
  const logit = HAPE_LOGIT_INTERCEPT + HAPE_LOGIT_SLOPE * total_points;
  return 1 / (1 + Math.exp(-logit));
}

/**
 * Estimate HAPE risk via the Suona 2023 nomogram (with SpO₂).
 *
 * @param age_years Age in years (study inclusion criterion: ≥14 years)
 * @param spo2_percent Peripheral oxygen saturation at altitude (%)
 * @param transport_mode 'plane' | 'train' | 'vehicle'
 * @param fatigue Patient reports fatigue
 * @param cough Patient reports cough of any kind (auto-set if sputum=true)
 * @param sputum Patient reports coughing sputum (white or pink, foamy)
 */
export function hapeRiskSuona2023(
  age_years: number,
  spo2_percent: number,
  transport_mode: TransportMode,
  fatigue: boolean,
  cough: boolean,
  sputum: boolean
): HAPERiskResult {
  if (age_years < 14) {
    throw new Error('age_years must be >= 14 (study inclusion criterion)');
  }
  if (!(spo2_percent > 0 && spo2_percent <= 100)) {
    throw new Error('spo2_percent must be in (0, 100]');
  }
  const mode = transport_mode.trim().toLowerCase() as TransportMode;
  if (mode !== 'plane' && mode !== 'train' && mode !== 'vehicle') {
    throw new Error("transport_mode must be 'plane', 'train', or 'vehicle'");
  }

  // If sputum is reported, cough is implied.
  const coughEffective = sputum ? true : cough;

  const ageCat = hapeAgeCategory(age_years);
  const ptsAge = HAPE_AGE_POINTS[ageCat];
  const ptsTransport = HAPE_TRANSPORT_POINTS[mode];
  const ptsFatigue = HAPE_FATIGUE_POINTS[String(fatigue) as 'true' | 'false'];
  const ptsCough = HAPE_COUGH_POINTS[String(coughEffective) as 'true' | 'false'];
  const ptsSputum = HAPE_SPUTUM_POINTS[String(sputum) as 'true' | 'false'];
  const ptsSpo2 = hapeSpo2ToPoints(spo2_percent);

  const totalPoints = ptsAge + ptsTransport + ptsFatigue + ptsCough + ptsSputum + ptsSpo2;
  const probability = Math.max(0, Math.min(1, hapePointsToProbability(totalPoints)));

  return { probability, total_points: totalPoints, model_used: 'with_spo2' };
}

// Utility functions
export function metersToFeet(meters: number): number {
  return meters / 0.3048;
}

export function feetToMeters(feet: number): number {
  return feet * 0.3048;
}

export function pascalToMmHg(pa: number): number {
  return pa / 133.322;
}

export function mmHgToPascal(mmHg: number): number {
  return mmHg * 133.322;
}
